tests/disk/test_cli.py

```diff
@@ -0,0 +1,573 @@
+
+        result = self.cli._get_removable_devices(disks)
+        assert len(result) == 0
+
+    @mock.patch(
+        "subprocess.Popen", side_effect=subprocess.CalledProcessError(1, "Popen")
+    )
+    def test_get_connected_devices_error(self, mocked_subprocess):
+
+        with pytest.raises(ExportException):
+            self.cli.get_connected_devices()
+
+    @mock.patch("subprocess.check_output", return_value=_SAMPLE_OUTPUT_NO_PART)
+    def test_get_partitioned_device_no_partition(self, mocked_call):
+        assert (
+            self.cli.get_partitioned_device(_DEFAULT_USB_DEVICE) == _DEFAULT_USB_DEVICE
+        )
+
+    @mock.patch("subprocess.check_output", return_value=_SAMPLE_OUTPUT_ONE_PART)
+    def test_get_partitioned_device_one_partition(self, mocked_call):
+        assert (
+            self.cli.get_partitioned_device(_DEFAULT_USB_DEVICE)
+            == _DEFAULT_USB_DEVICE + "1"
+        )
+
+    @mock.patch("subprocess.check_output", return_value=_SAMPLE_OUTPUT_MULTI_PART)
+    def test_get_partitioned_device_multi_partition(self, mocked_call):
+
+        with pytest.raises(ExportException) as ex:
+            self.cli.get_partitioned_device(_SAMPLE_OUTPUT_MULTI_PART)
+
+        assert ex.value.sdstatus is Status.INVALID_DEVICE_DETECTED
+
+    @mock.patch("subprocess.check_output", return_value=None)
+    def test_get_partitioned_device_lsblk_error(self, mocked_subprocess):
+        with pytest.raises(ExportException) as ex:
+            self.cli.get_partitioned_device(_SAMPLE_OUTPUT_ONE_PART)
+
+        assert ex.value.sdstatus is Status.DEVICE_ERROR
+
+    @mock.patch(
+        "subprocess.check_output",
+        side_effect=subprocess.CalledProcessError(1, "check_output"),
+    )
+    def test_get_partitioned_device_multi_partition_error(self, mocked_call):
+
+        # Make sure we wrap CalledProcessError and throw our own exception
+        with pytest.raises(ExportException) as ex:
+            self.cli.get_partitioned_device(_DEFAULT_USB_DEVICE)
+
+        assert ex.value.sdstatus is Status.DEVICE_ERROR
+
+    @mock.patch("subprocess.check_call", return_value=0)
+    def test_is_luks_volume_true(self, mocked_call):
+
+        # `sudo cryptsetup isLuks` returns 0 if true
+        assert self.cli.is_luks_volume(_SAMPLE_OUTPUT_ONE_PART)
+
+    @mock.patch(
+        "subprocess.check_call",
+        side_effect=subprocess.CalledProcessError(1, "check_call"),
+    )
+    def test_is_luks_volume_false(self, mocked_subprocess):
+
+        # `sudo cryptsetup isLuks` returns 1 if false; CalledProcessError is thrown
+        assert not self.cli.is_luks_volume(_SAMPLE_OUTPUT_ONE_PART)
+
+    @mock.patch("subprocess.check_output", return_value=_SAMPLE_LUKS_HEADER)
+    def test__get_luks_name_from_headers(self, mocked_subprocess):
+        result = self.cli._get_luks_name_from_headers(_DEFAULT_USB_DEVICE)
+
+        assert result is not None and result.split("-")[
+            1
+        ] in _SAMPLE_LUKS_HEADER.decode("utf8")
```

Comment:
Aren't the cases that test `self.cli._get_luks_name_from_headers(...)` not menaingfully covered by the tests on `self.cli.is_luks_volume(...)`?

If not: is there a way to test this though specific setup before acting on `self.cli.is_luks_volume(...)`?


Where my questions are coming from: I think that testing private methods is a smell. Since we are not meant to be invoking them from outside the class, I would expect their effects to be visible when calling public methods, and those _are_ tested in the scope of this unit test file.

It might seems like adding tests, even if redundant, doesn't hurt. I would argue that's not the case for a few reasons:

- if the specific cases being tested through private methods are not tested through the public methods that would invoke them, then the behavior of those public methods is not actually fully specified (documentation gap)
- when refactoring, having to make decisions on modifying or removing tests is an additional burden, that is unnecessary if all the visible (public) effects of the private method were tested elsewhere anyway
- test code is code, and there are less bug and less maintenance with code that's not there, so I wouldn't keep redundancy for the sake of non-meaningful test coverage

Now, what if the behavior being tested is difficult to set up around a public method, and is not tested otherwise? 
Then I'd argue that the private methods at play could very likely be extracted from the class and injected as a dependency. That way they would not be private anymore, for once, they would get their own unit tests, as a consequence, and what made them especially meaningful would be captured in the naming of the new class.

Reply:
Probably falls into the same category than https://github.com/freedomofpress/securedrop-export/pull/105#discussion_r1067553656